tracing-subscriber/src/filter/layer_filters/combinator.rs

```diff
@@ -0,0 +1,278 @@
+    /// // that will enable the `INFO` level and lower for spans and events with
+    /// // `my_crate` targets:
+    /// let filter = combinator::And::new(my_crate, LevelFilter::INFO);
+    ///
+    /// // If a span or event *doesn't* have a target beginning with
+    /// // `my_crate`, enable it if it has the `WARN` level or lower:
+    /// // let filter = combinator::Or::new(filter, LevelFilter::WARN);
+    ///
+    /// tracing_subscriber::registry()
+    ///     .with(tracing_subscriber::fmt::layer().with_filter(filter))
+    ///     .init();
+    /// ```
+    ///
+    /// [`Filter`]: crate::layer::Filter
+    pub const fn new(a: A, b: B) -> Self {
+        Self(a, b)
+    }
+}
+
+impl<S, A, B> Filter<S> for Or<A, B>
+where
+    A: Filter<S>,
+    B: Filter<S>,
+{
+    #[inline]
+    fn enabled(&self, meta: &Metadata<'_>, cx: &Context<'_, S>) -> bool {
+        self.0.enabled(meta, cx) || self.1.enabled(meta, cx)
+    }
+
+    fn callsite_enabled(&self, meta: &'static Metadata<'static>) -> Interest {
+        let a = self.0.callsite_enabled(meta);
+        let b = self.1.callsite_enabled(meta);
+
+        // If either filter will always enable the span or event, return `always`.
+        if a.is_always() || b.is_always() {
+            return Interest::always();
+        }
+
+        // Okay, if either filter will sometimes enable the span or event,
+        // return `sometimes`.
+        if a.is_sometimes() || b.is_sometimes() {
+            return Interest::sometimes();
+        }
+
+        debug_assert!(
+            a.is_never() && b.is_never(),
+            "if neither filter was `always` or `sometimes`, both must be `never` (a={:?}; b={:?})",
+            a,
+            b,
+        );
+        Interest::never()
+    }
+
+    fn max_level_hint(&self) -> Option<LevelFilter> {
+        // If either hint is `None`, return `None`. Otherwise, return the less restrictive.
+        Some(cmp::max(self.0.max_level_hint()?, self.1.max_level_hint()?))
+    }
+}
+
+// === impl Not ===
+
+impl<A> Not<A> {
+    /// Inverts the result of a [`Filter`].
+    ///
+    /// If the wrapped filter would enable a span or event, it will be disabled. If
+    /// it would disable a span or event, that span or event will be enabled.
+    ///
+    /// [`Filter`]: crate::layer::Filter
+    pub const fn new(a: A) -> Self {
+        Self(a)
+    }
+}
+
+impl<S, A> Filter<S> for Not<A>
+where
+    A: Filter<S>,
+{
+    #[inline]
+    fn enabled(&self, meta: &Metadata<'_>, cx: &Context<'_, S>) -> bool {
+        !self.0.enabled(meta, cx)
+    }
+
+    fn callsite_enabled(&self, meta: &'static Metadata<'static>) -> Interest {
+        match self.0.callsite_enabled(meta) {
+            i if i.is_always() => Interest::never(),
+            i if i.is_never() => Interest::always(),
+            _ => Interest::sometimes(),
+        }
+    }
+
+    fn max_level_hint(&self) -> Option<LevelFilter> {
+        // TODO(eliza): figure this out???
```

Comment:
this is the main reason this PR is still a draft

Reply:
jeez, yeah, that really fucks with my brain. i'd maybe return `None` for now

Reply:
The reason this is confusing is that I don't think we can simply return the "opposite" of the inner filter's hint, since the filter might be enabling things based on a particular characteristic other than levels, and we might still enable things of the same  level. I think maybe we should just return `None` and have this break caching...though honestly, I'm not totally sure how necessary the `Not` filter even is...